### PR TITLE
New toolchain: Updated "platform.txt"  to use the xpack gcc 9…

### PR DESCRIPTION
**Pull request by danieltr3s**
New toolchain: Updated "platform.txt"  to use the xpack gcc 9 toolchain; Added an extra option to choose between different runtime libraries in "boards.txt".

Issue: https://github.com/rogerclarkmelbourne/Arduino_STM32/issues/786

### CODE REVIEW

**Comment by stevstrong** [comment]
If you add `--specs=nano.specs` to the platform.txt, then you can remove  it from boards.txt, I think.

**Comment by fpistm** [comment]
> 
> 
> If you add `--specs=nano.specs` to the platform.txt, then you can remove it from boards.txt, I think.

No, in platform.txt this is the default value

**Comment by stevstrong** [comment]
I understand that it is the default value, but the same is repeated 3 times in boards.txt. What is the point to repeat it?

**Comment by fpistm** [comment]
it is not repeated , it is the value for `build.flags.ldspecs`
Depending of the selected menu then `build.flags.ldspecs` is set to the proper value. There is no concatenation, it is simply overidden.

**Comment by stevstrong** [comment]
If the default value is anyway overwritten by the boards.txt options, then it makes no sense to make a default value, just define it for each option in part.

**Comment by fpistm** [comment]
No for default option this is the one used and avoid to set it in the boards.txt:
`discovery_f407.menu.rtlib.nano=Newlib Nano (default)`

if you remove the one from platform.txt, then you will have to add
`discovery_f407.menu.rtlib.nano.build.flags.ldspecs=--specs=nano.specs`

for each entry...

**Comment by stevstrong** [comment]
As far as I can see, there are 5 new entries.
The nano-specs is already there for 3 entries, the first entry being the default, it should contain the same, and the last ld.specs entry then can be removed, if you do not add it into platform.txt.
Or am I missing something?

**Comment by fpistm** [comment]
> ?

Well I don't really understand what you would mean, anyway there is several other possibilities, 
This one is the one I use because it is the default option so it make sense to have it in the platform.txt for better readability.

File: STM32F1/platform.txt [hunk 1]
```diff
@@ -14,7 +14,7 @@ compiler.warning_flags.all=-Wall -Wextra -DDEBUG_LEVEL=DEBUG_ALL
 
 # compiler variables
 # ----------------------
-compiler.path={runtime.tools.arm-none-eabi-gcc.path}/bin/
+compiler.path={runtime.tools.xpack-arm-none-eabi-gcc-9.2.1-1.1.path}/bin/
```
[the remark this answers is not in the document]
Comment (danieltr3s) [reply STM32F1/platform.txt:17]: This path is generated and managed by the Arduino IDE in the same way as it is with the current toolchain.
When you install a package on the Arduino IDE, the JSON file describes all the boards and what toolchain each one uses.
Here's a partial of the STM32 Cores JSON file describing the toolchain and where the IDE should download it from:

image

Arduino IDE will then reference these tools by the "**name**" and will create a folder with that name inside the tools folder for that package. Then it will create another folder with the "**version**" as the name (so you can have different versions of the tools) and that's where the IDE will unpack the tools described in the JSON file.
Note that this JSON can be quite hard to digest as it typically describes all the versions of the package and each description is very, very long.

The _Wiki_ entry for **Installation** states (second bullet):
_Run the IDE, and on the Tools menu, select the Boards manager, and install the Arduino SAM boards (Cortex-M3) from the list of available boards. You must do this step, it installs the arm-none-eabi-g++ toolchain!_
And for this new toolchain, this step needs to change to **STM32 Cores** instead of **Arduino SAM boards**.

**Comment by danieltr3s** [comment]
> If you add `--specs=nano.specs` to the platform.txt, then you can remove it from boards.txt, I think.

No, because that will force that option for all compilations and you wouldn't be able to not use it.

> If the default value is anyway overwritten by the boards.txt options, then it makes no sense to make a default value, just define it for each option in part in boards.txt.

No. That's incorrect. The default value set in "**platforms.txt**" is only overridden in **"boards.txt"** on the options other than the _default_. If you notice, all the defaults in **"boards.txt"** do not set any compilation option. And as so, IDE will use what is defined in  "**platforms.txt**".

In order to not use the _nano.specs_ option (or any other for that matter) you have a menu entry definition that clears the compiler option. As @fpistm already said, the IDE does not do any concatenation, it simple overrides any value if one is provided in **"boards.txt"** or **"platform.local.txt"**.

The menu option definitions in **"boards.txt"** are read the following way:
[BoardName].menu.[menu option name].[option name] = [Display Text]  _-> This starts a new Option entry in the menu._
[BoardName].menu.[menu option name].[option name].build.flags.[name of the compile flag desired] = [value for the flag] _-> This sets a compile flag_
.... (repeat for any other compile flag)

The scope of the option ends when another "Option Name" is defined.
If you don't have any lines below that "Start of a new option", then no changes will be made to the build definitions in "**platforms.txt**".

To add new menus you have to add them at the beginning of the file:

menu.[name of menu] = [Display Text]

File: STM32F1/platform.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevstrong) [reply STM32F1/platform.txt:17]: You want to say that we need to install the STM32 core in order to use this PR?
I do not think that it is a viable solution, as this core should be independent form other cores.
This core is usually not installed using the Arduino package installer, but downloaded and extracted from ZIP.

**Comment by stevstrong** [comment]
Thank you for the explanation, I already know how the txt files work.

I just wanted to point out that you added 5 new RTL options for each board:
rtlib.nano, rtlib.nanofp, rtlib.nanofs, rtlib.nanofps and rtlib.full.

4 of these options overriding the default `flags.ldspecs` as newly specified in platform.txt, the first having the default value (no overriding ldspecs).

I think this can be done in the following way:
- revert the added `build.flags.ldspecs=--specs=nano.specs` in platform.txt
- adding `.rtlib.nano.build.flags.ldspecs=--specs=nano.specs` for each board in boards.txt - this will explicitly set the ldspecs for the first option Newlib Nano (default)
- remove `.rtlib.full.build.flags.ldspecs=` for each board in boards.txt -  this will keep the default old ldspecs option.

This way you leave the default value of ldspecs for Newlib standard and explicitly set it for `.rtlib.nano`.

Do you now understand what I mean?

**Comment by danieltr3s** [comment]
> * .rtlib.nano.build.flags.ldspecs=--specs=nano.specs

Perfectly, @stevstrong . 
I just fail to see the benefits of such approach. In fact, that would mean that the default options of the menus in boards.txt wouldn't match with the defaults declared in platform.txt which can lead to painful hours of cross checking between files (talking with experience :D).

What do you think?

File: STM32F1/platform.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (danieltr3s) [reply STM32F1/platform.txt:17]: However you still have to use the Arduino package installer to install the SAMD boards, as stated in the Wiki. Otherwise you won't have the required toolchain. It's just a matter of installing a different package to use a different toolchain.

image

I'm investigating a way of having both toolchains available that I think will please you.

[the remark this answers is not in the document]
Comment (danieltr3s) [reply STM32F1/platform.txt:17]: Update: Just pushed a new commit that adds an option to change between xPack and the old arm-none-eabi-gcc .
I think this will make it easier to see the gains between using the new tools and the old ones and improve on that.

[the remark this answers is not in the document]
Comment (stevstrong) [reply STM32F1/platform.txt:17]: I do not think that installing an additional **board package** (SAM boards) can be made equivalent to installing an additional **core**.

I think the users can download the new xpack separately and copy it to the old arm eabi directory, without installing a complete new core.

[the remark this answers is not in the document]
Comment (danieltr3s) [reply STM32F1/platform.txt:17]: It is exactly the same, @stevstrong. When someone installs the SAM boards package, that person is actually installing an additional full Core. Your suggestion is to (if someone wants to use more recent and better tools) potentially break their installation of the other Cores and Boards. Doesn't seem right.

**Comment by stevstrong** [comment]
> 
> In fact, that would mean that the default options of the menus in boards.txt wouldn't match with the defaults declared in platform.txt which can lead to painful hours of cross checking between files (talking with experience :D).

AFAIK, there is no default option in boards.txt. The options in boards.txt are used to override the default option in platform.txt.

File: STM32F1/platform.txt [hunk 2]
```diff
@@ -45,7 +45,7 @@ build.cpu_flags=
 build.hs_flag=
 build.upload_flags=
 build.flags.optimize=-Os
-build.flags.ldspecs=
+build.flags.ldspecs=--specs=nano.specs
```
Comment (stevstrong) [inline STM32F1/platform.txt:48]: I have the feeling that this option will increase the code site instead of reducing it.
I downloaded the STM32DUINO package in Arduino boards manager ( which took 1 GB space !!!) and changed the platform.txt to use the xpack instead of old EABI package.
Without nano-specs the code size was 14,704 bytes, with it was 23,556 bytes....

So am have huge doubts whether this PR really brings any advantage to this core.

Comment (fpistm) [reply STM32F1/platform.txt:48]: > I have the feeling that this option will increase the code size instead of reducing it.

No. If this is the case then you have a mistake with the config.

> STM32DUINO package

when you told this you talk about which one ?

Comment (stevstrong) [reply STM32F1/platform.txt:48]: I mean STM32DUINO package the official ST package.

Regarding the code size, what kind of configuration do you mean? I just tested the platform.txt by adding the mentioned flag, and the outcome was as I wrote, an increase in size.

File: STM32F1/platform.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevstrong) [reply STM32F1/platform.txt:17]: The SAM board package is ~250MB large, the STM32DUINO package is ~1 GB large, also 4 times larger. Alone form this point of view it makes a huge difference to have on or the other package...

File: STM32F1/platform.txt (hunk 2 above)
Comment (fpistm) [reply STM32F1/platform.txt:48]: OK, I misunderstood your point. You download it to get the gcc and try with your core 😉 

Anyway, I've already test in the past and did not get those results. Maybe some libc functions used by the core are huge with the newlib nano. For example assert.h usage is huge. Also usage of float for printf and scanf can increase size anyway I'm really surprise of your result as by default this should be less.

File: STM32F1/platform.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (fpistm) [reply STM32F1/platform.txt:17]: > Alone form this point of view it makes a huge difference to have on or the other package...

Yes CMSIS and arm gcc are huge.

File: STM32F1/platform.txt (hunk 2 above)
Comment (fpistm) [reply STM32F1/platform.txt:48]: Here my test with the master of the Arduino_STM432 repo:

# Blink example

### Default config: gcc 7-2017q4 and no nanospecs
Sketch uses 15836 bytes (14%) of program storage space. Maximum is 110592 bytes.
Global variables use 3176 bytes (18%) of dynamic memory, leaving 14232 bytes for local variables. Maximum is 17408 bytes.

### Config: gcc 7-2017q4 and nanospecs
Sketch uses 14276 bytes (12%) of program storage space. Maximum is 110592 bytes.
Global variables use 2160 bytes (12%) of dynamic memory, leaving 15248 bytes for local variables. Maximum is 17408 bytes.

### Config: xpaxk gcc 9.2.1-1.1 and no nanospecs
Sketch uses 15736 bytes (14%) of program storage space. Maximum is 110592 bytes.
Global variables use 3176 bytes (18%) of dynamic memory, leaving 14232 bytes for local variables. Maximum is 17408 bytes.

### Config: xpaxk gcc 9.2.1-1.1 and nanospecs
Sketch uses 14272 bytes (12%) of program storage space. Maximum is 110592 bytes.
Global variables use 2160 bytes (12%) of dynamic memory, leaving 15248 bytes for local variables. Maximum is 17408 bytes.

So there is a benefits with nanospecs but with the toolchain it is not so evident. But I guess with a real sketch using more feature it would be better.

Anyway, I really don't understand how you get your results.

Comment (stevstrong) [reply STM32F1/platform.txt:48]: OK, I have to correct my information, previously I did the test with my repo, now using Roger's repo:

arm-none-eabi-gcc:
```
Sketch uses 15384 bytes (23%) of program storage space. Maximum is 65536 bytes.
Global variables use 3176 bytes (15%) of dynamic memory, leaving 17304 bytes for local variables. Maximum is 20480 bytes.
```

xpack-arm-none-eabi-gcc-9.2.1-1.1:
```
Sketch uses 15316 bytes (23%) of program storage space. Maximum is 65536 bytes.
Global variables use 3176 bytes (15%) of dynamic memory, leaving 17304 bytes for local variables. Maximum is 20480 bytes.
```

arm-none-eabi-gcc + nano.specs:
```
Sketch uses 13940 bytes (21%) of program storage space. Maximum is 65536 bytes.
Global variables use 2064 bytes (10%) of dynamic memory, leaving 18416 bytes for local variables. Maximum is 20480 bytes.
```

xpack-arm-none-eabi-gcc-9.2.1-1.1 +nano.specs:
```
Sketch uses 13936 bytes (21%) of program storage space. Maximum is 65536 bytes.
Global variables use 2064 bytes (10%) of dynamic memory, leaving 18416 bytes for local variables. Maximum is 20480 bytes.
```

As a conclusion, while using the nano,specs brings visible code size improvement, it does not bring to much additionally when using the xpack-arm-none-eabi-gcc-9.2.1-1.1.

So at this point I am still not very motivated to load ~1GB of data (the official core including XPACK) to reduce only a few bytes in the code size.

Comment (rogerclarkmelbourne) [reply STM32F1/platform.txt:48]: I agree with Steve, the space saving by using xpack-arm-none-eabi-gcc-9.2.1-1.1. is minimal, and there would need to be a huge amount of manual regression testing to confirm that xpack-arm-none-eabi-gcc-9.2.1-1.1. does not cause any runtime problems with code, including things like RTOS compatibility, and IRQ, DMR , USB compatibility etc

Re:Nanospecs
I don't know the details of what library code gets removed when Nanospecs is used, normally you can't reduce the binary size by 10%+ with loosing some functionality.
The core may not be using the functionality, but some existing application code may be.

Comment (fpistm) [reply STM32F1/platform.txt:48]: simply add the newlib nano specs as an option and kept standard newlib as default.
